File: frontend/src/api/allergies.ts
import { api } from "./client";

export interface Allergy {
  id: number;
  ingredient_name: string;
  notes: string | null;
}

export function listAllergies(): Promise<Allergy[]> {
  return api.get<Allergy[]>("/api/allergies");
}

export function addAllergy(ingredientName: string, notes?: string): Promise<Allergy> {
  return api.post<Allergy>("/api/allergies", {
    ingredient_name: ingredientName,
    notes: notes ?? null,
  });
}

export function deleteAllergy(id: number): Promise<void> {
  return api.delete(`/api/allergies/${id}`);
}

File: backend/app/controllers/allergies.py
from fastapi import APIRouter, Depends

from app.dependencies import get_allergy_service
from app.dto.allergy_dto import AllergyDtoIn, AllergyDtoOut
from app.services.allergy_service import AllergyService

router = APIRouter(prefix="/allergies", tags=["allergies"])


@router.get("")
def list_allergies(service: AllergyService = Depends(get_allergy_service)) -> list[AllergyDtoOut]:
    return [AllergyDtoOut.from_domain(allergy) for allergy in service.list_allergies()]


@router.post("", status_code=201)
def add_allergy(
    dto: AllergyDtoIn,
    service: AllergyService = Depends(get_allergy_service),
) -> AllergyDtoOut:
    return AllergyDtoOut.from_domain(service.add_allergy(dto.to_domain()))


@router.delete("/{allergy_id}", status_code=204)
def delete_allergy(
    allergy_id: int,
    service: AllergyService = Depends(get_allergy_service),
) -> None:
    service.remove_allergy(allergy_id)
